src/xopen/__init__.py

```diff
@@ -614,9 +652,42 @@ def _detect_format_from_extension(filename: Union[str, bytes]) -> Optional[str]:
     return None
 
 
+def _file_or_path_to_binary_stream(
+    file_or_path: FileOrPath, binary_mode: str
+) -> Tuple[BinaryIO, bool]:
+    if binary_mode not in ("rb", "wb", "ab"):
+        raise AssertionError()
+    if file_or_path == "-":
+        return _open_stdin_or_out(binary_mode), False
+    if isinstance(file_or_path, (str, bytes)) or hasattr(file_or_path, "__fspath__"):
+        return open(os.fspath(file_or_path), binary_mode), True  # type: ignore
+    if isinstance(file_or_path, (io.BufferedReader, io.BufferedWriter)):
+        return file_or_path, False
+    if isinstance(file_or_path, io.TextIOWrapper):
+        return file_or_path.buffer, False
+    if isinstance(file_or_path, io.IOBase) and not hasattr(file_or_path, "encoding"):
+        # Text files have encoding attributes. This file is binary:
+        return file_or_path, False
+    else:
+        raise TypeError(
+            f"Unsupported type for {file_or_path}, "
```

Comment:
Hm, I wonder if this isn’t too restrictive. For the sake of duck typing, would it maybe be enough to not have most of the above cases and just check whether the object has a `read()` method instead?

Reply:
Agreed. I opted for "readinto" as that is done in dnaio IIRC for binary files.